ote_cli/ote_cli/utils/hpo.py

```diff
@@ -0,0 +1,660 @@
+"""
+Utils for HPO with hpopt
+"""
+
+import builtins
+import collections
+import importlib
+import multiprocessing
+import os
+import pickle
+import shutil
+import subprocess
+import sys
+import time
+from os import path as osp
+from pathlib import Path
+from typing import Optional
+
+import numpy
+import torch
+import yaml
+from ote_sdk.configuration.helper import create
+from ote_sdk.entities.model import ModelEntity, ModelStatus
+from ote_sdk.entities.model_template import TaskType
+from ote_sdk.entities.subset import Subset
+from ote_sdk.entities.task_environment import TaskEnvironment
+from ote_sdk.entities.train_parameters import TrainParameters, UpdateProgressCallback
+
+from ote_cli.datasets import get_dataset_class
+from ote_cli.utils.importing import get_impl_class
+from ote_cli.utils.io import generate_label_schema
+
+try:
+    import hpopt
+except ImportError:
+    hpopt = None
+
+
+def check_hpopt_available():
+    """Check whether hpopt is avaiable"""
+
+    if hpopt is None:
+        return False
+    return True
+
+
+def run_hpo(args, environment, dataset, task_type):
+    """Update the environment with better hyper-parameters found by HPO"""
+    hyper_parameters = None
+
+    if check_hpopt_available():
+        if task_type not in {TaskType.CLASSIFICATION, TaskType.DETECTION}:
+            print(
+                "Currently supported task types are classification and detection."
+                f"{task_type} is not supported yet."
+            )
+            return None
+
+        dataset_paths = {
+            "train_ann_file": args.train_ann_files,
+            "train_data_root": args.train_data_roots,
+            "val_ann_file": args.val_ann_files,
+            "val_data_root": args.val_data_roots,
+        }
+
+        hpo_save_path = os.path.abspath(
+            os.path.join(os.path.dirname(args.save_model_to), "hpo")
+        )
+        hpo = HpoManager(
+            environment, dataset, dataset_paths, args.hpo_time_ratio, hpo_save_path
+        )
+        hyper_parameters = hpo.run()
+
+        environment.set_hyper_parameters(hyper_parameters)
+
+        if args.load_weights:
```

Comment:
What is it needed for?

Reply:
When using pre-trained weights, it is needed to modify hyper-parameters in environment.model.configurable_parameters to work properly.